aten/src/ATen/core/dispatch/DispatchKeyExtractor.h

```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include <cstdint>
+#include <ATen/core/function_schema.h>
+#include <ATen/core/jit_type.h>
+#include <c10/core/TensorTypeSet.h>
+
+namespace c10 {
+
+/**
+ * An instance of DispatchKeyExtractor knows how to get a dispatch key given
+ * a list of arguments for an operator call. The instance is specific for
+ * a certain operator as different operators have different ways to extract
+ * the dispatch key (e.g. different numbers of arguments).
+ */
+struct DispatchKeyExtractor final {
+public:
+  static DispatchKeyExtractor make(const FunctionSchema& schema) {
+    bool is_valid = false;
+    for (size_t i = 0; i < schema.arguments().size(); ++i) {
+      const auto& type = schema.arguments()[i].type();
+      if (type->isSubtypeOf(TensorType::get())) {
+        is_valid = true;
+        break;
+      }
+      if (type->isSubtypeOf(ListType::ofTensors())) {
+        is_valid = true;
+        break;
+      }
+    }
+
+    return DispatchKeyExtractor(schema.arguments().size(), is_valid);
+  }
+
+  c10::optional<TensorTypeId> getDispatchKeyBoxed(const Stack* stack) const {
+    if (C10_UNLIKELY(!is_valid_)) {
+      return c10::nullopt;
+    }
+
+    TensorTypeSet ts;
+    for (const auto& ivalue : torch::jit::last(*stack, num_args_)) {
+      if (C10_LIKELY(ivalue.isTensor())) {
+        // NB: Take care not to introduce a refcount bump (there's
+        // no safe toTensorRef method, alas)
+        ts = ts | ivalue.unsafeToTensorImpl()->type_set();
+      } else if (C10_UNLIKELY(ivalue.isTensorList())) {
+        for (const auto& tensor : ivalue.toTensorListRef()) {
+          ts = ts | tensor.type_set();
+        }
+      }
+    }
+    // TODO: Don't use legacy extractor; blocked on c10 understanding variable
+    return c10::legacyExtractTypeId(ts);
+  }
+
+  template<class... Args>
+  c10::optional<TensorTypeId> getDispatchKeyUnboxed(const Args&... args) const {
+    AT_ASSERT(sizeof...(args) == num_args_, "Wrong number of arguments. Schema says ", num_args_, " but was called with ", sizeof...(args));
```

Comment:
this check is in the hot inlined path, isn't it? did you check that it doesn't blow up the codesize? also, do we need this check at all? if we assume that unboxed version is called by codegen only, it's probably safe to assume that it's ok. Or we can make it a DCHECK or something

Reply:
I'll remove the check.